### PR TITLE
[TIR] Add pass to check for out of bounds memory access

### PR DESCRIPTION
**Pull request by tkonolige**
This is a conservative static analysis that checks to see if any out of bounds array access occurs. It is not enabled by default.

@Lunderberg @AndrewZhaoLuo @jwfromm 

cc @gigiblender

### CODE REVIEW

File: src/tir/analysis/oob_checker.cc [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  template <class T>
+  void CheckBounds(const T* node, size_t i) {
+    auto ind_bounds = analyzer_.int_set(node->indices[i]);
+    auto shape_bounds = analyzer_.int_set(node->buffer->shape[i]);
+    // Only show an error if we can prove that the access occurs out of bounds.
+    // In some cases we may not be able to prove that the access is in or out
+    // of bounds and we would like to ignore these cases.
+    if (analyzer_.CanProve(ind_bounds.max() >= shape_bounds.min()) ||
```
Comment (Lunderberg) [inline src/tir/analysis/oob_checker.cc:96]: Do we need to use the `int_set` analyzer here?  Why not `analyzer_.CanProve(node->indices[i] < 0 || node->indices[i] >= node->buffer->shape[i])`.

Comment (tkonolige) [reply src/tir/analysis/oob_checker.cc:104]: Good question! `CanProve` uses universal quantification — are all indices out of bounds. What we really want is existential quantification — is any index out of bounds. We could check `!analyzer_.CanProve(node->indices[i] >= 0 && node->indices[i] < node->buffer->shape[I])` but then we run into the issue of having some access patterns that are all inbounds but not provably so. I wanted this analysis to be conservative (no false positives) so we can't use this check either. The remaining solution is to check if the bounds of the index lie outside the bounds for the shape.

Comment (Lunderberg) [reply src/tir/analysis/oob_checker.cc:104]: Thank you, and that makes sense!

**Comment by AndrewZhaoLuo** [comment]
I'll take a look over the weekend

**Review by Lunderberg — APPROVED** [review]
LGTM!

**Review by AndrewZhaoLuo — APPROVED** [review]

**Pull request merged**